File: server/models/packet_server_scene_movement.go
package models

import (
	"encoding/json"
	"github.com/pkg/errors"
	"unlock/constants"
)

type PacketServerSceneMovement struct {
	Packet
	Coordinates *Coordinates       `json:"coordinates"`
	ClientUuid  string             `json:"clientUuid"`
	Motion      *Motion            `json:"motion"`
	SceneKey    constants.SceneKey `json:"sceneKey"`
}

func NewPacketServerSceneMovement(coordinates *Coordinates, clientUuid string, motion *Motion, sceneKey constants.SceneKey) *PacketServerSceneMovement {
	return &PacketServerSceneMovement{
		Packet: Packet{
			Label: constants.PacketServerSceneMovement,
		},
		Coordinates: coordinates,
		ClientUuid:  clientUuid,
		Motion:      motion,
		SceneKey:    sceneKey,
	}
}

func (packet *PacketServerSceneMovement) Send(lobby *Lobby) (err error) {
	payload, err := json.Marshal(packet)
	if err != nil {
		return errors.WithStack(err)
	}

	lobby.Broadcast <- payload

	return nil
}

File: client/src/models/packets/packet-server-scene-movement.js
import PacketLabels from '../../constants/packet-labels'

export default class PacketServerSceneMovement {

  // noinspection DuplicatedCode
  constructor(packet) {
    this.label = PacketLabels.SERVER_SCENE_MOVEMENT
    this.clientUuid = packet.clientUuid || null
    this.x = packet.coordinates?.x || null
    this.y = packet.coordinates?.y || null
    this.vx = packet.coordinates?.vx || null
    this.vy = packet.coordinates?.vy || null
    this.ax = packet.coordinates?.ax || null
    this.ay = packet.coordinates?.ay || null
    this.r = packet.coordinates?.r || null
    this.direction = packet.motion?.direction || null
    this.jumping = packet.motion?.jumping || null
    this.walking = packet.motion?.walking || null
    this.sceneKey = packet.sceneKey || null
  }

  receive(sceneKey, callback) {
    if (this.sceneKey !== sceneKey) {
      return
    }

    if (typeof callback === 'function') {
      callback(this)
    }
  }

}
